server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -124,74 +176,109 @@ public void init() {
      * period not lapsing, the result will be of type {@link AutoShardingType#COOLDOWN_PREVENTED_INCREASE} or
      * {@link AutoShardingType#COOLDOWN_PREVENTED_INCREASE} with the remaining cooldown configured and the number of shards that should
      * be configured for the data stream once the remaining cooldown lapses as the target number of shards.
      *
-     * The NOT_APPLICABLE type result will report a cooldown period of TimeValue.MAX_VALUE.
+     * <p>The NOT_APPLICABLE type result will report a cooldown period of TimeValue.MAX_VALUE.
      *
-     * The NO_CHANGE_REQUIRED type will potentially report the remaining cooldown always report a cool down period of TimeValue.ZERO (as
+     * <p>The NO_CHANGE_REQUIRED type will potentially report the remaining cooldown always report a cool down period of TimeValue.ZERO (as
      * there'll be no new auto sharding event)
      */
-    public AutoShardingResult calculate(ProjectState state, DataStream dataStream, @Nullable Double writeIndexLoad) {
+    public AutoShardingResult calculate(ProjectState state, DataStream dataStream, IndexStats writeIndexStats) {
         if (isAutoShardingEnabled == false) {
-            logger.debug("Data stream auto sharding service is not enabled.");
+            logger.debug("Data stream auto-sharding service is not enabled.");
             return NOT_APPLICABLE_RESULT;
         }
 
         if (dataStreamExcludePatterns.stream().anyMatch(pattern -> Regex.simpleMatch(pattern, dataStream.getName()))) {
             logger.debug(
-                "Data stream [{}] is excluded from auto sharding via the [{}] setting",
+                "Data stream [{}] is excluded from auto-sharding via the [{}] setting",
                 dataStream.getName(),
                 DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING.getKey()
             );
             return NOT_APPLICABLE_RESULT;
         }
 
-        if (writeIndexLoad == null) {
+        if (writeIndexStats == null) {
             logger.debug(
-                "Data stream auto sharding service cannot compute the optimal number of shards for data stream [{}] as the write index "
-                    + "load is not available",
+                "Data stream auto-sharding service cannot compute the optimal number of shards for data stream [{}] as the write index "
+                    + "stats are not available",
                 dataStream.getName()
             );
             return NOT_APPLICABLE_RESULT;
         }
-        return innerCalculate(state.metadata(), dataStream, writeIndexLoad, nowSupplier);
+
+        double writeIndexLoad = sumLoadMetrics(writeIndexStats, IndexingStats.Stats::getWriteLoad);
+        double writeIndexRecentLoad = sumLoadMetrics(writeIndexStats, IndexingStats.Stats::getRecentWriteLoad);
+        double writeIndexPeakLoad = sumLoadMetrics(writeIndexStats, IndexingStats.Stats::getPeakWriteLoad);
```

Comment:
Nit: This is nice and readable, but if performance becomes an issue we could consider calculating them in one loop.  I do not think this is a critical path (executed all the time etc), so this might be ok.